Add emojis:generate_borders Rake task

leo60228 opened the pull request:
cc @ThibG 

See #13769, #13766

leo60228 committed:
Add emojis:generate_borders Rake task

leo60228 commented:
Oops, didn't mean to modify the lockfile...

ThibG commented on lib/tasks/emojis.rake:
```diff
@@ -1,5 +1,45 @@
 
+def gen_border(codepoint)
+  input = Rails.root.join('public', 'emoji', "#{codepoint}.svg")
+  dest = Rails.root.join('public', 'emoji', "#{codepoint}_border.svg")
+  doc = File.open(input) { |f| Nokogiri::XML(f) }
+  svg = doc.at_css("svg")
+  if svg.key?("viewBox")
+    viewBox = svg["viewBox"].split(" ").map { |s| s.to_i }
+    viewBox[0] -= 2
+    viewBox[1] -= 2
+    viewBox[2] += 4
+    viewBox[3] += 4
+    svg["viewBox"] = viewBox.join(" ")
+  end
+  g = Nokogiri::XML::Node.new "g", doc
+  for elem in doc.css("svg > *")
+    border_elem = elem.dup
+
+    if border_elem.key?("fill")
+      border_elem.delete("fill")
+    end
+    border_elem["stroke"] = "white"
+
+    style = ""
+    if border_elem.key?("style")
+      style = border_elem["style"]
+    end
+    old_width = "0px"
+    if border_elem.key?("stroke-width")
+      old_width = border_elem["stroke-width"]
+    end
+    style += " stroke-width: calc(#{old_width} + 4px)"
+    border_elem["style"] = style.strip
+
+    g.add_child(border_elem)
+  end
+  svg.prepend_child(g)
+  File.write(dest, doc.to_xml)
+  puts "Wrote bordered #{codepoint}.svg to #{dest}!"
+end
```
CodeClimate complains about a bunch of styling issues.

ThibG commented on public/emoji/.gitignore:
```diff
@@ -0,0 +1 @@
+*_border.svg
```
My idea was to have the generated files in the repo, like we may do with the lock files, or the `emoji_map.json`.

ThibG commented on lib/tasks/emojis.rake:
```diff
@@ -55,4 +95,11 @@ namespace :emojis do
   end
+
+  desc 'Generate emoji variants with white borders'
+  task :generate_borders, [] do |task, args|
+    for cc in args.extras
+      gen_border cc
+    end
+  end
```
My idea was that the unicode characters to generate files from would be hardcoded in here.

ClearlyClaire reviewed (changes requested):
I have made a few requests for changes, but I can take over if you prefer, as I have a good idea what I want to do with the front-end part in particular.

leo60228 committed:
Address review

leo60228 commented:
Done.

leo60228 committed:
Border all dark emoji

ThibG commented on lib/tasks/emojis.rake:
```diff
@@ -55,4 +89,12 @@ namespace :emojis do
+
+  desc 'Generate emoji variants with white borders'
+  task :generate_borders do
+    codepoints = %w(1f3b1 1f41c 1f519 26ab 1f5a4 2b1b 25fe 25fc 2712 25aa 1f4a3 1f3b3 1f464 1f465 1f4f7 1f4f8 2663 27b0 1f4b1 1f576 2734 1f50c 1f51a 1f482-200d-2640-fe0f 1f482-1f3fb-200d-2640-fe0f 1f482-1f3fc-200d-2640-fe0f 1f482-1f3fd-200d-2640-fe0f 1f482-1f3fe-200d-2640-fe0f 1f482-1f3ff-200d-2640-fe0f 1f4fd 1f373 1f98d 1f482 1f482-1f3fb 1f482-1f3fc 1f482-1f3fd 1f482-1f3fe 1f482-1f3ff 2714 2797  ...
+    codepoints.each do |cc|
+      gen_border cc
+    end
```
oh, I was going to suggest something like

```ruby
src = Rails.root.join('app', 'javascript', 'mastodon', 'features', 'emoji', 'emoji_map.json')
    emojis = '🎱🐜⚫🖤⬛◼️◾◼️✒️▪️💣🎳📷📸♣️🕶️✴️🔌💂‍♀️📽️🍳🦍💂🔪🕳️🕹️🕋🖊️🖋️💂‍♂️🎤🎓🎥🎼♠️🎩🦃📼📹🎮🐃🏴👽⚾🐔☁️💨🕊️👀🍥👻🐐❕❔⛸️🌩️🔊🔇📃🌧️🐏🍚🍙🐓🐑💀☠️🌨️🔉🔈💬💭🏐🏳️⚪⬜◽◻️▫️'

    map = Oj.load(File.read(src))

    emojis.each_grapheme_cluster do |emoji|
      gen_border map[emoji]
    end
```

leo60228 commented:
This approach causes artifacting with sharp internal corners. I'm going to see if this approach can be combined with the filter approach.

ThibG commented:
Thanks again for your work, applying this on top of #13772 and adding a few front-end changes, I get the following for the dark theme:
![image](https://user-images.githubusercontent.com/384364/82159905-fe28fb80-9891-11ea-8546-8eb6471be33b.png)

And this for the light theme:
![image](https://user-images.githubusercontent.com/384364/82159977-5b24b180-9892-11ea-8fe7-e916eae6d7ef.png)

It looks pretty much the same on Firefox 76, Chromium 81 and Midori 7, so that is really good! On the very old Firefox 47, the borders do not look as good, but they're still here and make the whole thing legible.

leo60228 committed:
Combine stroke with filter to reduce artifacting

leo60228 commented:
I had very bad artifacting with emoji like the chicken. This tweaked filter is able to hide artifacting without affecting much else, and while seamlessly degrading on browsers without filter support.

leo60228 commented:
The camera with flash emoji looks somewhat odd as one of the corners, despite being internal, is still included in the outline. I'm going to try and tweak it by hand, it shouldn't affect much anyway.

ThibG commented:
Oh I hadn't noticed you went away from filters to do it purely with the stroke size. Are there browsers without filter support? If not, is there any downside in using only the filter, and not changing the stroke width?

leo60228 committed:
Cleanup Camera with Flash

leo60228 commented:
IE9 is the main one. I'm not sure if Mastodon supports IE9. There's no real downside, but it'll require some more tweaking.

ThibG commented:
Ok. As you wish. Working with stroke width might also benefit from using `stroke-linejoin`

leo60228 commented:
There's tons of small artifacts when using only the filter:
![image](https://user-images.githubusercontent.com/8355305/82160672-f6069700-9864-11ea-8509-5a747614323a.png)
Not sure how noticable they'll be zoomed out, though.

leo60228 committed:
Add stroke-linejoin="round"

The previous filter and tweaks were effectively a poor imitation of it.
There are no artifacts for any dark emoji now!

ThibG commented:
Yeah, only using the filter isn't that good. I wonder if `stroke-linejoin: round` doesn't solve all the issues you had which required the filter.
Also I think I get why the borders weren't working as well on Firefox 47: it doesn't handle calc(). Maybe you could use “4px” in all cases, as it seems Twemoji simply doesn't use strokes at all, and thus it will always be 0px.

leo60228 committed:
Set stroke-width using property

This fixes old versions of Firefox.

leo60228 committed:
Store emoji in string instead of array

leo60228 commented:
Thank you for the `stroke-linejoin` tip! It works perfectly.

leo60228 commented:
I got the *opposite* error when running `rubocop` locally...

leo60228 committed:
Use separate arguments for each path segment

leo60228 commented:
Are you making a separate PR for frontend changes or should I keep this as a draft?

ThibG commented on lib/tasks/emojis.rake:
```diff
@@ -1,5 +1,36 @@
+    g.add_child(border_elem)
+  end
+  svg.prepend_child(g)
+  svg['style'] = 'background: black;'
```
Oops, please remove that :)

ThibG commented:
I think I'll make the front-end changes in another PR. You can mark this PR as ready for review.

leo60228 committed:
Remove "background: black;"

ThibG commented:
Please refresh the generated SVGs as well
